apps/base-docs/docs/building-with-base/guides/onchain-generative-nfts.md

```diff
@@ -0,0 +1,876 @@
+- Cloud: 6 KB
+- Sky: 802 bytes
+
+If you don't have the tools to do this, you can find these files here: [TODO LINK]
+
+## Contract Architecture
+
+You'll need to build and deploy a number of contracts for this project. They'll be organized in this architecture:
+
+![Architecture](../../../assets/images/onchain-generative-nfts/architecture.png)
+
+Deploying this many contracts will have a cost associated with it, but once they're deployed, this contract will cost the same as any other NFT contract. Remember, `pure` and `view` functions called outside the blockchain don't cost any gas. This means that we can use multiple contracts to assemble a relatively large graphic without additional costs!
+
+## Building the Contracts
+
+Create a new project using the toolkit of your choice, and add a contract called `LandSeaSkyNFT`. Import OpenZeppelin's ERC-721, inherit from it, and set it up with the constructor, an mint function, and a counter to keep track of the token ID:
+
+```solidity
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.20;
+
+import "hardhat/console.sol";
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+
+contract LandSeaSkyNFT is ERC721 {
+    uint public counter;
+
+    constructor() ERC721("Land, Sea, and Sky", "LSS") {}
+
+    function mint() public {
+        counter++;
+        _safeMint(msg.sender, counter);
+    }
+}
+```
+
+### Overriding the `_baseURI()` Function
+
+Normally, you'd override `_baseURI()` with the base URL for the location you select to keep your NFT metadata. This could be a website, IPFS folder, or many other possible locations.
+
+Since this contract will be generating the .json file directly, instead set it to indicate this to the browser:
+
+```solidity
+function _baseURI() internal pure override returns (string memory) {
+  return "data:application/json;base64,";
+}
+```
+
+### Importing the Base64 Library
+
+As indicated above, you'll be returning the json metadata in [Base64] format. OpenZeppelin has a utility contract to do this. You'll also need the `Strings` library. Go ahead and import them:
+
+```solidity
+import "@openzeppelin/contracts/utils/Base64.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+```
+
+:::info
+
+Base64 allows the transport of binary data over the web in a reliable way. It is not a compression algorithm, and actually increased the data size by a 4/3 ratio.
+
+:::
+
+### Planning the override for the `tokenURI()` Function
+
+Next, set up your `tokenURI` function override. You'll need to write some other contracts to make this work, but you can write most of the code and stub out a plan for the rest to:
+
+- Check and ensure the token ID exists
+- Compile the json metadata for the token, including:
+  - The `"name"` of the NFT
+  - A `"description"`
+  - The `"image"`
+- Base64 encode the above, combine it with the `_baseURI` and return it.
+
+```solidity
+function tokenURI(uint _tokenId) public view override returns (string memory) {
```

Comment:
Should `_tokenId` be `tokenId` here, based on the variable used within the body?

Reply:
Other way around, but yes.  Good catch!